### DISCUSSION TITLE
ADR - GitOps for App Platform Components

### DISCUSSION THREAD

**Discussion by dudash** [dudash/psychdocs#3] [adr]
### Details

We need to install several components that are "app platform" level. They aren't part of the common development workflow but are more like lego pieces needed for the app microservices. Essentially, middleware pieces of the system (message bus, SSO, API gateway, etc). This ADR is about how to install and configure them. Specifically, should GitOps be used for this purpose.

### Decision

We assume an empty day 1 OpenShift cluster is available. And then we will use ArgoCD to bootstrap install and manage several platform components. Including:
* ArgoCD itself
* Istio service mesh
* Keycloak SSO
* API gateway

### Rationale

These components are build by 3rd parties and just need to be installed and configured. They change rarely so there will be a bit of unnecessary overhead to do GitOps on them. So it almost seems counter intutive, maybe we should just `kubctl apply` or `helm install` and forget about it. However I do like the aspect that they **shouldn't** change so if they do we can have history in a PR. And if that change is unintentional/imperative, we have GitOps automatically put them back to the declared configuration/state (or warn us about it) - reliably consistent.

### Consequences

Pros: consistency, GitOps dashboards, history of changes in PRs, `argocd rollback` possible

Cons: complexity of learning ArgoCD to manage, if things go wrong it's harder to imperatively make changes

### Choose status below. If deprecated, indicate why. If superseded, include a link to the new ADR

Proposed

### Authors

@dudash

**Comment by dudash** [comment]
as a comment of the con about imperative actions: Red Hat validated patterns suggest cronjobs + ansible as a potential solution https://validatedpatterns.io/patterns/multicloud-gitops/mcg-imperative-actions/

"There is currently no way within Argo CD to apply an imperative action against a cluster. However, you can declaratively apply changes to a cluster using kubernetes cronjob resources. Customers can use their Ansible playbooks to take action against a cluster if necessary."

However, I find this to be a somewhat complicated workaround

**Comment by jkeam** [comment]
My two cents:

Reminds me of us hacking on this a few days ago, when we needed to turn off ArgoCD's sync in order to perform some upgrades.  I think in an ideal world, we would use ArgoCD for everything including the platform components, but I think the added complexity isn't worth it.  

When we want to be able to make dynamic changes quickly to get the cluster to a working state for instance, instead of pushing our changes through ArgoCD, we have to be able to debug it and incrementally change things right on the cluster.  I'm coming at it from a pragmatic's lens as opposed to a dogmatic approach.

Now this doesn't necessarily mean no IaC (Infra as code).  The yaml's you are using you would still keep in source control.  But you might have needed to run some extra `oc` commands to destroy some random pods that wouldn't die or CRDs that need to be cleaned out on a particularly problematic cluster.  And maybe you don't have to do this in any other cluster.  For that kind of stuff, I wouldn't check that into any IaC repo.

I think ArgoCD really shines for apps.  I think this is a solid use case and I think ArgoCD's aggressive remediation/restoration back to your IaC repo definition is good for those pesky developers that tend to just connect to the cluster and quickly "fix" things.  The GitOps approach here forces them to be deliberate around their changes, potentially requiring a lead or someone to approve the change, communicates the change to the rest of the team, and keeps a beautiful history with comments around these changes.

In conclusion, right tool for the right job and I think the nail for the hammer that is ArgoCD are apps.  Of course I'm open to any dissenting opinion and would love discourse on this.

**Reply by jkeam** [reply]
That being said, this looks pretty nice.  https://cloud.redhat.com/blog/creating-azure-red-hat-openshift-and-red-hat-openshift-services-on-aws-clusters-declaratively

So who knows, maybe I'm wrong on everything?

**Reply by jkeam** [reply]
And of course: https://github.com/redhat-na-ssa/openshift-dev-workflows-gitops-cluster-bootstrap

**Reply by jkeam** [reply]
Maybe Operators are just another workload?  So then it's ArgoCD's job?

**Reply by gbengataylor** [reply]
plus this https://validatedpatterns.io/patterns/multicloud-gitops/#overview-architecture

**Reply by gbengataylor** [reply]
oops, dudash references that

**Comment by gbengataylor** [comment]
see an example 

> plus this https://validatedpatterns.io/patterns/multicloud-gitops/#overview-architecture

**Reply by gbengataylor** [reply]
and i just spoke to Cory - we might be looking to  adopt this for our team next year https://github.com/redhat-na-ssa/demo-ai-gitops-catalog

**Reply by codekow** [reply]
This is my starting point when saving time for operator and certain demo elements:
https://github.com/redhat-cop/gitops-catalog

If the solution is not in there it ends up in:
https://github.com/redhat-na-ssa/demo-ai-gitops-catalog

Or it's own repo with references to the other repo with `kustomize`
